extensions/amp-viewer-integration/0.1/findtext.js

```diff
@@ -0,0 +1,451 @@
+  }
+
+  /**
+   * @return {number}
+   */
+  get char() {
+    return this.node.wholeText[this.offset];
+  }
+}
+
+/**
+ * TextRange represents a text range.
+ * Exported for test only.
+ */
+export class TextRange {
+  /**
+   * @param {!TextPos} start
+   * @param {!TextPos} end
+   */
+  constructor(start, end) {
+    this.start = start;
+    this.end = end;
+  }
+}
+
+const skipCharRe = /[,.\s\u2022()]/;
+
+/**
+ * @param {string} c input char to normalize.
+ * @return {string}
+ */
+const normalizeChar = function(c) {
+  if (c == '\u2019' || c == '\u2018') {
+    return '\'';
+  }
+  if (c == '\u201c' || c == '\u201d') {
+    return '"';
+  }
+  return c.toLowerCase();
+};
+
+/**
+ * @param {string} s input string to normalize.
+ * @return {string}
+ */
+export const normalizeString = function(s) {
+  const buf = [];
+  for (let i = 0; i < s.length; i++) {
+    const c = s[i];
+    if (skipCharRe.test(c)) {
+      continue;
+    }
+    buf.push(normalizeChar(c));
+  }
+  return buf.join('');
+};
+
+/**
+ * findSentences find sentences from node and returns a list of TextRange.
+ * @param {!Node} node
+ * @param {!Array<string>} sentences
+ * @return {?Array<!TextRange>}
+ */
+export const findSentences = function(node, sentences) {
+  const scanner = new TextScanner(node);
+  const matches = [];
+  for (let senIdx = 0; senIdx < sentences.length; senIdx++) {
+    const sen = normalizeString(sentences[senIdx]);
+    if (!sen) {
+      continue;
+    }
+    // BM-algorithm with bad-character rules.
+    const skipTable = {};
+    for (let i = 0; i < sen.length; i++) {
+      const c = sen[i];
+      skipTable[c] = sen.length - 1 - i;
+    }
+    const buf = new CircularBuffer(sen.length);
+    let index = -1;
+    let nextIndex = sen.length - 1;
+    while (true) {
+      const pos = scanner.next();
+      if (pos == null) {
+        // mismatch
+        return null;
+      }
+      if (pos == posDomDelimiter || skipCharRe.test(pos.char)) {
+        continue;
+      }
+      buf.add(pos);
+      index++;
+      if (index < nextIndex) {
+        continue;
+      }
+      let ok = true;
+      for (let j = 0; j < sen.length; j++) {
+        const c = normalizeChar(buf.get(sen.length - j - 1).char);
+        if (sen[sen.length - 1 - j] == c) {
+          continue;
+        }
+        ok = false;
+        let skip = skipTable[c];
+        if (skip == null) {
+          skip = sen.length;
+        }
+        skip -= j;
+        if (skip < 1) {
+          skip = 1;
+        }
+        nextIndex += skip;
+        break;
+      }
+      if (ok) {
+        const endPos = buf.get(sen.length - 1);
+        matches.push(new TextRange(
+            buf.get(0), new TextPos(endPos.node, endPos.offset + 1)));
+        break;
+      }
+    }
+  }
+  return matches.length > 0 ? matches : null;
+};
+
+/**
+ * @param {!Array<!TextRange>} ranges
+ * @return {!Array<!Element>} A list of marked nodes.
+ */
+export const markTextRangeList = function(ranges) {
```

Comment:
Common AMP style for module-level functions is to use `function name(args)` if possible. Please update signatures.

Reply:
Done. I also updated other functions I added.